core/GlobalState.cc

```diff
@@ -833,11 +828,17 @@ SymbolRef GlobalState::enterSymbol(Loc loc, SymbolRef owner, NameRef name, u4 fl
     return ret;
 }
 
-SymbolRef GlobalState::enterClassSymbol(Loc loc, SymbolRef owner, NameRef name) {
+SymbolRef GlobalState::enterClassSymbol(Loc loc, SymbolRef owner, NameRef name, bool createSingleton) {
     ENFORCE(!owner.exists() || // used when entering entirely syntehtic classes
             owner.data(*this)->isClassOrModule());
     ENFORCE(name.data(*this)->isClassName(*this));
-    return enterSymbol(loc, owner, name, Symbol::Flags::CLASS_OR_MODULE);
+    auto symsUsed = symbolsUsed();
+    SymbolRef sym = enterSymbol(loc, owner, name, Symbol::Flags::CLASS_OR_MODULE);
+    if (createSingleton && (sym._id >= symsUsed)) {
+        // New symbol (and thus has no singleton class)
+        sym.data(*this)->defineSingletonClass(*this);
+    }
+    return sym;
```

Comment:
I might have misunderstood dmitry's suggestion, but I don't think this alone gets us the most amount of performance as we can get. I interpreted his suggest to mean that we change `enterSymbol` itself to re-use some intermediate data structures to directly create a singleton class when entering a CLASS_OR_MODULE symbol kind.

I'm assuming you're checking the performance of this, so if it makes it faster I'm happy to land, but I just wanted to tag @DarkDimius in case this was not the implementation he had in mind.

Reply:
> we change enterSymbol itself to re-use some intermediate data structures to directly create a singleton class when entering a CLASS_OR_MODULE symbol kind

That change would remove another hash lookup, but then we'd have to re-implement `defineSingleton` without `enterClassSymbol` within `enterSymbol`. And then we'd still have to have a second code path, implemented on `Symbol`, that lazily instantiates singleton classes. I tried removing `singletonClass` entirely -- removing the option to lazily instantiate -- but sometimes we request a `singletonClass` of a singleton class (assuming pretty printed `<class: Foo>` indicates the name of a singleton class).

Reply:
Yup, my suggestion was to make `enterClassSymbol` not rely on `enterSymbol` and to do two lookups at once.
This gets us half-way-there